### DISCUSSION TITLE
iPXE on Rock64

### DISCUSSION THREAD

**Discussion by CobraFlow** [ipxe/ipxe#1096] [General]
Hi guys,
I am having trouble trying to get even the simplest iPXE configuration to work. The board has a Realtek RTL8211F network card.
I thought that I would start off loading the snp.efi image from the TFTP server and moving on from there.
* I have flashed the board with Ayufan's uboot loader.
* Configured a TFTP server
* Built snp.efi using 
`make bin-arm64-efi/snp.efi`
* Configured dnsmasq to return /snp.efi
The uboot loads the file but exits with a return code of 2.
![Rock64_iPXE](https://github.com/ipxe/ipxe/assets/3617910/5c5ffdd7-6b09-42c9-8741-179c4ae8165a)

Any help would be greatly appreciated...

**Comment by NiKiZe** [comment]
Could you get any efi application working? Such as efi shell?

**Reply by CobraFlow** [reply]
This was my initial foray into iPXE. I have tried nothing else. Is there another application that I could try first...?

**Reply by NiKiZe** [reply]
> efi shell

**Reply by CobraFlow** [reply]
What is efi shell. Is it something that I boot off SD card or can I use uboot and tftp server??

**Reply by CobraFlow** [reply]
Ok, long delay in getting back to this.

I successfully load grubaa64.efi with no changes.

I am building ipxe with 

`make CROSS=aarch64-linux-gnu- bin-arm64-efi/ipxe.efi -j4`

The r=2 is the problem.

**Comment by CobraFlow** [comment]
I have compiled iPXE with DEBUG=efi_init and rerun. I get these messages...

![iPXE error 1](https://github.com/ipxe/ipxe/assets/3617910/db9d8eac-98ed-4022-878d-36259ebc0cab)

**Reply by NiKiZe** [reply]
If you have text, post it as text, preferably in a code block, follow any ipxe.org urls that is included in any fatal messages.

**Comment by CobraFlow** [comment]
Hi guys. I'm picking this up again. I have just cloned the latest repo and am building on the arm machine itself.

The messages are a little different but the outcome is the same...
EFI handle 0xfcf040f0 systab 0xfefa4888
EFI protocol ConsoleControl is at 0xfef7f188
EFI does not provide protocol ef9fc172-a1b2-4693-b327-6d32fc416042
EFI does not provide protocol AcpiTable
EFI protocol 8b843e20-8132-4852-90cc-551a4e4a7f1c is at 0xfef7f1a0
EFI does not provide configuration table Acpi10
EFI configuration table Fdt is at 0x8200000
EFI does not provide configuration table Smbios3
EFI configuration table Smbios is at 0xfbee3000
EFI image base address 0x2000000
EFI could not get loaded image's device path: Error 0x7f2f2082 (https://ipxe.org/7f2f2082)## Application terminated, r = 2

Any help would be much appreciated!

Thank you...

**Reply by NiKiZe** [reply]
This text seems wrong, the error url is broken. Is this actually coped text, or maybe a OCR of image with errors? Also use code blocks to make the text more readable.
Do you have any issue open with Ayufan's bootloader or whatever the EFI implementation is?

**Reply by CobraFlow** [reply]
Hi NiKiZe,

The text is exactly as is. I cloned the repo and built it. 

No, no problems with the bootloader. I can do 'normal' pxe loads of operating systems. 

The initial testing of this above consists of the DHCP server returning ipxe.efi as the bootfile. Nothing fancy at all!!

Not sure what I should build though. I'm building on the arm machine itself... What should my build command look like? 


Cheers

**Reply by NiKiZe** [reply]
Follow this link from your text: https://ipxe.org/7f2f2082 isn't found, so something is of.

You clearly have an interoperability issue between the EFI firmware and iPXE. If the firmware lacks support for features that is expected to exist by iPXE, than this is an issue that is better solved there. In this case, I assume that EFI firmware doesn't adhere to what most other firmwares do provide, either not providing a boot source path at all, or not making it possible to open that path.

Try building iPXE maybe with `DEBUG=efi_utils,efi_init` but if you follow the url in the message (non corrupt version that is) you get that suggestion.

In earlier image you had https://ipxe.org/7f39e082 which makes more sense.

**Reply by CobraFlow** [reply]
If I build directly from clone using the master branch I get exactly the output above with the incorrect URL. If I checkout the latest tag, v1.21.1, I get the following build error...

`  [BUILD] bin-arm64-efi/fcp.o
In function ‘fc_xchg_respond’,
    inlined from ‘fc_port_deliver’ at net/fc.c:899:10:
net/fc.c:775:31: error: array subscript 0 is outside array bounds of ‘struct fc_responder[0]’ [-Werror=array-bounds]
  775 |                 if ( responder->type == type ) {
      |                      ~~~~~~~~~^~~~~~
In file included from include/ipxe/efi/efi.h:66,
                 from include/ipxe/errno/efi.h:26,
                 from include/errno.h:109,
                 from net/fc.c:30:
net/fc.c: In function ‘fc_port_deliver’:
include/ipxe/tables.h:250:39: note: object ‘__table_entries’ of size 0
  250 |         static __table_type ( table ) __table_entries[0]                \
      |                                       ^~~~~~~~~~~~~~~
include/ipxe/tables.h:271:30: note: in expansion of macro ‘__table_entries’
  271 | #define table_start( table ) __table_entries ( table, 00 )
      |                              ^~~~~~~~~~~~~~~
include/ipxe/tables.h:359:25: note: in expansion of macro ‘table_start’
  359 |         for ( pointer = table_start ( table ) ;                         \
      |                         ^~~~~~~~~~~
net/fc.c:774:9: note: in expansion of macro ‘for_each_table_entry’
  774 |         for_each_table_entry ( responder, FC_RESPONDERS ) {
      |         ^~~~~~~~~~~~~~~~~~~~
In function ‘fc_xchg_respond’,
    inlined from ‘fc_port_deliver’ at net/fc.c:899:10:
net/fc.c:776:46: error: array subscript 0 is outside array bounds of ‘struct fc_responder[0]’ [-Werror=array-bounds]
  776 |                         if ( ( rc = responder->respond ( &xchg->ulp, port,
      |                                     ~~~~~~~~~^~~~~~~~~
net/fc.c: In function ‘fc_port_deliver’:
include/ipxe/tables.h:250:39: note: object ‘__table_entries’ of size 0
  250 |         static __table_type ( table ) __table_entries[0]                \
      |                                       ^~~~~~~~~~~~~~~
include/ipxe/tables.h:271:30: note: in expansion of macro ‘__table_entries’
  271 | #define table_start( table ) __table_entries ( table, 00 )
      |                              ^~~~~~~~~~~~~~~
include/ipxe/tables.h:359:25: note: in expansion of macro ‘table_start’
  359 |         for ( pointer = table_start ( table ) ;                         \
      |                         ^~~~~~~~~~~
net/fc.c:774:9: note: in expansion of macro ‘for_each_table_entry’
  774 |         for_each_table_entry ( responder, FC_RESPONDERS ) {
      |         ^~~~~~~~~~~~~~~~~~~~
  [BUILD] bin-arm64-efi/fragment.o
`

I was not expecting that!

**Comment by CobraFlow** [comment]
Thanks Christian, I'll give it a go tomorrow and report back.

🙏

*Lewis Henderson*

Director
CobraFlow Limited

M:0788 7788 436
www.cobraflow.com

**Comment by mcb30** [comment]
> EFI could not get loaded image's device path: Error 0x7f2f2082 (https://ipxe.org/7f2f2082)## Application terminated, r = 2

Breaking down the expected flow behind this error message:

* EFI (in this case, U-Boot) passes an `EFI_HANDLE` to the iPXE entry point, which is the handle of the loaded iPXE image itself.
* iPXE opens the `EFI_LOADED_IMAGE_PROTOCOL` instance on this image handle
* iPXE retrieves the `DeviceHandle` from this protocol instance, which is the EFI handle of the device from which iPXE was loaded.
* iPXE opens the `EFI_DEVICE_PATH_PROTOCOL` instance on this device handle.  This is the point that is shown as failing by the debug messages.

The issue is that U-Boot is either providing an invalid `DeviceHandle`, or is failing to provide an `EFI_DEVICE_PATH_PROTOCOL` for this device handle.

This is not a general upstream U-Boot problem: I have personally loaded iPXE from U-Boot many times.  Since you mentioned that you "flashed the board with Ayufan's uboot loader", I'm guessing that this is a U-Boot fork rather than the upstream.  Try using the upstream U-Boot tree instead.

**Reply by CobraFlow** [reply]
Thank you for your patience. I'll give that a try. I'm very slow at this as I have a real job and this is only moving forward in my own time!!
